lib/Controller/SettingsController.php

```diff
@@ -148,8 +166,74 @@ public function impersonate($userId) {
 				'app' => 'impersonate',
 			]
 		);
+		$this->issueWarning($userId, $currentUser->getUID());
+		$this->sendEmail($userId, $currentUser->getUID());
 		$this->userSession->setUser($user);
 		return new JSONResponse();
 	}
+
+	/**
+	 * Issues the warning by creating a notification
+	 *
+	 * @param string $userId
+	 * @param string $impersonator
+	 */
+	protected function issueWarning($userId, $impersonator) {
+		$notification = $this->notificationManager->createNotification();
+		try {
+			$notification->setApp('impersonate')
+				->setObject('impersonation', $userId)
+				->setUser($userId)
+				->setDateTime(new \DateTime())
+				->setSubject('impersonate', ['impersonator' => $impersonator]);
+			$this->notificationManager->notify($notification);
+		} catch (\InvalidArgumentException $e) {
+			$this->logger->logException($e, ['app' => 'impersonate']);
+		}
+	}
+
+	/**
+	 * Send an email to the user
+	 *
+	 * @param string $userId
+	 * @param float $impersonator
+	 */
+	protected function sendEmail($userId, $impersonator) {
+		$user = $this->userManager->get($userId);
+		if (!$user instanceof IUser) {
+			return;
+		}
+
+		$email = $user->getEMailAddress();
+		if (!$email) {
+			return;
+		}
+
+		$lang = $this->config->getUserValue($userId, 'core', 'lang');
+		$l = $this->l10nFactory->get('impersonate', $lang);
+		$emailTemplate = $this->mailer->createEMailTemplate('impersonate.Notification', [
+			'impersonator' => $impersonator,
+			'userId' => $user->getUID()
+		]);
+
+		$emailTemplate->addHeader();
+		$emailTemplate->addHeading($l->t('A user impersonating as you'), false);
```

Comment:
```suggestion
		$emailTemplate->addHeading($l->t('A user impersonated you on Nextcloud'), false);
```

Reply:
Ideally even use the instance's name from Theming, but I don't know how
cc @nickvergessen

Reply:
Emails are not going to be sent, and adding Nextcloud is bad.

Reply:
I'm removing the email functionality then? @nickvergessen

Reply:
So the notification will be indismissable for the impersonator?
Can the user at least opt in to emails via their notification settings?